pkg/sdn/plugin/sdn-cni-plugin/openshift-sdn.go

```diff
@@ -0,0 +1,611 @@
+	mapArgs, err := gatherCniArgs(args.Args)
+	if err != nil {
+		return err
+	}
+
+	// CNI doesn't yet have an UPDATE command so we fake it through CNI_ARGS.
+	// See https://github.com/containernetworking/cni/issues/89
+	if action, ok := mapArgs["OPENSHIFT_ACTION"]; ok && action == "UPDATE" {
+		return cmdUpdate(args)
+	}
+
+	podConfig, err := getPodConfig(args)
+	if err != nil {
+		return err
+	}
+
+	// Run IPAM so we can set up container veth
+	ipamConfig, err := getIPAMConfig(podConfig)
+	if err != nil {
+		return err
+	}
+
+	os.Setenv("CNI_ARGS", "")
+	result, err := ipam.ExecAdd("host-local", ipamConfig)
+	if err != nil {
+		return fmt.Errorf("failed to run CNI IPAM ADD: %v", err)
+	}
+	if result.IP4 == nil || result.IP4.IP.IP.To4() == nil {
+		return fmt.Errorf("failed to obtain IP address from CNI IPAM")
+	}
+
+	var hostVeth, contVeth netlink.Link
+	err = ns.WithNetNSPath(args.Netns, func(hostNS ns.NetNS) error {
+		hostVeth, contVeth, err = ip.SetupVeth(interfaceName, int(podConfig.mtu), hostNS)
+		if err != nil {
+			return fmt.Errorf("failed to create container veth: %v", err)
+		}
+		// refetch to get hardware address and other properties
+		contVeth, err = netlink.LinkByIndex(contVeth.Attrs().Index)
+		if err != nil {
+			return fmt.Errorf("failed to fetch container veth: %v", err)
+		}
+
+		// Clear out gateway to prevent ConfigureIface from adding the cluster
+		// subnet via the gateway
+		result.IP4.Gateway = nil
+		if err = ipam.ConfigureIface(interfaceName, result); err != nil {
+			return fmt.Errorf("failed to configure container IPAM: %v", err)
+		}
+
+		lo, err := netlink.LinkByName("lo")
+		if err == nil {
+			err = netlink.LinkSetUp(lo)
+		}
+		if err != nil {
+			return fmt.Errorf("failed to configure container loopback: %v", err)
+		}
+		return nil
+	})
+	if err != nil {
+		ipam.ExecDel("host-local", ipamConfig)
+		return err
+	}
+
+	if podConfig.wantMacvlan {
+		if err := addMacvlan(args.Netns); err != nil {
+			ipam.ExecDel("host-local", ipamConfig)
+			return err
+		}
+	}
+
+	contVethMac := contVeth.Attrs().HardwareAddr.String()
+	podIP := result.IP4.IP.IP.String()
+	out, err := exec.Command(sdnScript, setUpCmd, hostVeth.Attrs().Name, contVethMac, podIP, podConfig.vnid, podConfig.ingressBandwidth, podConfig.egressBandwidth).CombinedOutput()
+	if isScriptError(err) {
+		ipam.ExecDel("host-local", ipamConfig)
+		return fmt.Errorf("error running network setup script:\nhostVethName %s, contVethMac %s, podIP %s, podConfig %#v\n %s", hostVeth.Attrs().Name, contVethMac, podIP, podConfig, out)
+	} else if err != nil {
+		ipam.ExecDel("host-local", ipamConfig)
+		return err
+	}
+
+	// Push PodIP to apiserver; node process needs it for HostPort handling
+	updatePodIP(podConfig.masterKubeConfig, args.Args, podIP)
+
+	// Wait for node process to set up any pod hostports
+	if err := waitForPodStatus(podConfig.podStatusPath); err != nil {
+		ipam.ExecDel("host-local", ipamConfig)
+		return err
+	}
+
+	return result.Print()
```

Comment:
no, i mean, if result.Print() returns an error, then we'll return an error from cmdAdd(), presumably causing the pod to be scuttled, but we won't release its IP